### ISSUE TITLE
Run `beforeScenario` only once before a scenario

### ISSUE THREAD

**Issue by rvarbanov** [artilleryio/artillery#842]
I thought that I can use the `beforeScenario` to call my auth service to get an access token and use that for each call to the service I'm load testing. However, my `auth` function is called before each endpoint call. That is suboptimal because I don't want load test my auth service at the same time I load test all my other services.

Code example:
```yml
scenarios:
  - name: "Endpoint returns 200"
    beforeScenario: "auth"
    flow:
      - get:
          url: "/api/v1/endpoint"
          expect:
            - statusCode: 200
            - contentType: json
```

**Comment by AnkitGuhe** [comment]
@rvarbanov Wondering if you found the solution?

**Comment by joberdick-rcd** [comment]
Is there really not a way to do a one time call that excludes being ran by all virtual users?

**Comment by hassy** [comment]
`beforeScenario` is indeed meant to be run before each scenario execution (by each virtual user).

There is a way to run a call just once via a `before` section at the top level of the test definition. The contents of the section can be an arbitrary scenario, but any variables captured by that scenario will be available to all virtual users.

Quick example:

```yaml
config:
  target: https://artillery.io
  phases:
    - duration: 10
      arrivalRate: 1

before:
  flow:
    - log: "running a request before virtual users run"
    # Grab a random URL from the homepage:
    - get:
        url: "/"
        capture:
          - selector: "a"
            attr: "href"
            index: "random"
            as: "url"

scenarios:
  # All virtual users will make requests to the same URL
  - flow:
      - log: "URL is {{ url }}"
      - get:
          url: "{{ url }}"
```

**Comment by savitanagaraj** [comment]
> `beforeScenario` is indeed meant to be run before each scenario execution (by each virtual user).
> 
> There is a way to run a call just once via a `before` section at the top level of the test definition. The contents of the section can be an arbitrary scenario, but any variables captured by that scenario will be available to all virtual users.
> 
> Quick example:
> 
> ```yaml
> config:
>   target: https://artillery.io
>   phases:
>     - duration: 10
>       arrivalRate: 1
> 
> before:
>   flow:
>     - log: "running a request before virtual users run"
>     # Grab a random URL from the homepage:
>     - get:
>         url: "/"
>         capture:
>           - selector: "a"
>             attr: "href"
>             index: "random"
>             as: "url"
> 
> scenarios:
>   # All virtual users will make requests to the same URL
>   - flow:
>       - log: "URL is {{ url }}"
>       - get:
>           url: "{{ url }}"
> ```

@hassy I gave this a try. The exact same way. I had a post in the "before" section that could get me a bearer token, and I wanted to use that for the rest of my scenarios. I am getting an error. I have a bug logged for it #959 :
```
$ DEBUG=http:response artillery run Get_Token.yml
- TypeError: Cannot read property 'concat' of undefined
   at HttpEngine.createScenario (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\engine_http.js:91:23)
   at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:511:37
   at new Promise (<anonymous>)
   at handleBeforeRequests (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:501:10)
   at runner (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:132:25)
   at Runner.run (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\runner-sp.js:60:3)
   at done (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\commands\run.js:191:14)
   at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\node_modules\async\lib\async.js:52:16
   at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\node_modules\async\lib\async.js:1209:30
   at done (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\commands\run.js:267:14)
```
My code is similar to this:
```
config:
  target: 'https://url1'
  phases:
   - name: "Default"
     duration: 1
     arrivalRate: 2
  payload:
    - path: "userlogin.csv"
      fields:
        - user1
        - user2
        - password1
        - password2
      order: "sequence"
before:
 - flow:
    #Get bearer token
   - log: "Getting token"
   - post:
      url: "https://url2/oauth2/v2.0/token"
      json:
       username: 'XXXX'
       password: 'XXXX'
       grant_type: 'password'
       client_id: '12345'
       client_secret: 'abcde'
       scope: 'openid
      capture:
       json: "$.id_token"
       as: "id_token"
      expect:
        - statusCode: 200
scenarios:
 - flow:
   - get:
       url: "url3"
       headers:
         authorization: 'Bearer {{id_token}}'
         user-login: "XXXX"
         user-role: "admin"

**Comment by savitanagaraj** [comment]
Looks like the `before` block work only with json files and not yaml.

**Comment by AnkitGuhe** [comment]
That's right before block only works with JSON file!

**Comment by savitanagaraj** [comment]
@AnkitGuhe it works with yaml too! @hassy  responded to a bug I had logged. The contents of config.before should be an object rather than a list. That means, it should not be `- flow`, but `flow`. It works.

**Comment by AnkitGuhe** [comment]
I see, thanks for correcting @savitanagaraj, I had to rewrite all my tests in past and that was a pain. Appreciate it!

**Comment by MarcoArredondo** [comment]
Do you know if is there an afterScenario instruction too? I need to delete a parent element that I created in the beforeScenario section. I looked trhoug the documentation but I see nothing

`config:
  target: "https://api.todoist.com"
  phases:
    - duration: 10
      arrivalRate: 1
      name: "Phase 1"
  defaults: 
    headers:
      Authorization: Bearer XXX
      Content-Type: application/json
      X-Requested-Id: $(uuidgen)
before:
  flow:
    - log: "Creating a test project"
    - post: 
        url: /rest/v1/projects
        json:
          name: "Artillery Project"
        capture:
          json: "$.id"
          as: "projectID"
    - log: "Project ID to work with is {{ projectID }}"
after:
  flow:
    - log: "Deleting project {{projectID}}"
    - delete: 
        url: /rest/v1/projects/{{projectID}}
scenarios: 
  - flow:      
      - get: 
          url: /rest/v1/tasks
      - post:
          url: /rest/v1/tasks
          json: 
            content: "Artillery Test"
            project_id: "{{projectID}}"
          capture:
            json: "$.id"
            as: "taskID"
      - get: 
          url: /rest/v1/tasks/{{taskID}}
      - post:
          url: /rest/v1/tasks/{{taskID}}
          json: 
            content: "New task content"
      - post:
          url: /rest/v1/tasks/{{taskID}}/close
      - post:
          url: /rest/v1/tasks/{{taskID}}/reopen
      - delete:
          url: /rest/v1/tasks/{{taskID}}`

I did this, the before works correctly but the after seems to do nothing.

**Comment by Conrad6** [comment]
> > `beforeScenario` is indeed meant to be run before each scenario execution (by each virtual user).
> > There is a way to run a call just once via a `before` section at the top level of the test definition. The contents of the section can be an arbitrary scenario, but any variables captured by that scenario will be available to all virtual users.
> > Quick example:
> > ```yaml
> > config:
> >   target: https://artillery.io
> >   phases:
> >     - duration: 10
> >       arrivalRate: 1
> > 
> > before:
> >   flow:
> >     - log: "running a request before virtual users run"
> >     # Grab a random URL from the homepage:
> >     - get:
> >         url: "/"
> >         capture:
> >           - selector: "a"
> >             attr: "href"
> >             index: "random"
> >             as: "url"
> > 
> > scenarios:
> >   # All virtual users will make requests to the same URL
> >   - flow:
> >       - log: "URL is {{ url }}"
> >       - get:
> >           url: "{{ url }}"
> > ```
> 
> @hassy I gave this a try. The exact same way. I had a post in the "before" section that could get me a bearer token, and I wanted to use that for the rest of my scenarios. I am getting an error. I have a bug logged for it #959 :
> 
> ```
> $ DEBUG=http:response artillery run Get_Token.yml
> - TypeError: Cannot read property 'concat' of undefined
>    at HttpEngine.createScenario (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\engine_http.js:91:23)
>    at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:511:37
>    at new Promise (<anonymous>)
>    at handleBeforeRequests (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:501:10)
>    at runner (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\core\lib\runner.js:132:25)
>    at Runner.run (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\runner-sp.js:60:3)
>    at done (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\commands\run.js:191:14)
>    at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\node_modules\async\lib\async.js:52:16
>    at C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\node_modules\async\lib\async.js:1209:30
>    at done (C:\Users\snagaraj\AppData\Roaming\nvm\v12.20.1\node_modules\artillery\lib\commands\run.js:267:14)
> ```
> 
> My code is similar to this:
> 
> ```
> config:
>   target: 'https://url1'
>   phases:
>    - name: "Default"
>      duration: 1
>      arrivalRate: 2
>   payload:
>     - path: "userlogin.csv"
>       fields:
>         - user1
>         - user2
>         - password1
>         - password2
>       order: "sequence"
> before:
>  - flow:
>     #Get bearer token
>    - log: "Getting token"
>    - post:
>       url: "https://url2/oauth2/v2.0/token"
>       json:
>        username: 'XXXX'
>        password: 'XXXX'
>        grant_type: 'password'
>        client_id: '12345'
>        client_secret: 'abcde'
>        scope: 'openid
>       capture:
>        json: "$.id_token"
>        as: "id_token"
>       expect:
>         - statusCode: 200
> scenarios:
>  - flow:
>    - get:
>        url: "url3"
>        headers:
>          authorization: 'Bearer {{id_token}}'
>          user-login: "XXXX"
>          user-role: "admin"
> ```

Can the `beforeScenario` be used to modify other properties like the headers, url, etc?

**Comment by judaschwartz** [comment]
I am trying to login Before running my flow (I do not want the login requests to be part of the scenario results
I tried moving the login flow to the beforeScenario but that doesn't seem to run  
```
config:
  target: '{{ $processEnvironment.URL }}'
scenarios:
  - name: 'Warm Up'
    beforeScenario:
      flow:
        - post:
            url: '/login'
            json:
              username: 'admin'
              password: 'password'
            capture:
              - json: '$.user._id'
                as: 'userId'
    flow:
      - post:
          url: '/api/graphql'
          json:
            - operationName: 'apiRequest'
              query: 'query testQuery($userId: ID!) { testQuery: testGraphqlQuery(userId: $userId) { id name } } '
              variables:
                userId: '{{ userId }}'
```
Is there any way to achieve this?

**Comment by MickL** [comment]
@rvarbanov I think this issue can be closed as `before` and `after` are meant to be run only once per test as described [in the documentation](https://docs-nine-inky.vercel.app/docs/guides/guides/test-script-reference#before-and-after-sections).

**Comment by MickL** [comment]
@judaschwartz may you explain the downvote? Isnt that exactly what was requested?

**Comment by jackykwan-eventx** [comment]
> @judaschwartz may you explain the downvote? Isnt that exactly what was requested?

i think he need beforeScenario, not beforeTest

**Comment by ashokyamsani** [comment]
> `beforeScenario` is indeed meant to be run before each scenario execution (by each virtual user).
> 
> There is a way to run a call just once via a `before` section at the top level of the test definition. The contents of the section can be an arbitrary scenario, but any variables captured by that scenario will be available to all virtual users.
> 
> Quick example:
> 
> ```yaml
> config:
>   target: https://artillery.io
>   phases:
>     - duration: 10
>       arrivalRate: 1
> 
> before:
>   flow:
>     - log: "running a request before virtual users run"
>     # Grab a random URL from the homepage:
>     - get:
>         url: "/"
>         capture:
>           - selector: "a"
>             attr: "href"
>             index: "random"
>             as: "url"
> 
> scenarios:
>   # All virtual users will make requests to the same URL
>   - flow:
>       - log: "URL is {{ url }}"
>       - get:
>           url: "{{ url }}"
> ```

can we execute a function from processors to initialize page and browsers.
